### DISCUSSION TITLE
Upload issues with imagick + LiteSpeed on WordPress

### DISCUSSION THREAD

**Discussion by MrIverM** [ImageMagick/ImageMagick#2528] [Help]
Hi guys,

We just installed imagick on all our servers beacause WordPress have started displaying  "modul error" connected to imagick under "Site health", that says that imagick should be installed. Many of our clients have noticted this, and thus we have installed it.

However, imagick is not working well with LiteSpeed webserver, and our customers are now getting image upload-errors like described here:

https://networksandrants.com/2020/04/05/cant-upload-image-to-wordpress-due-to-large-size/?fbclid=IwAR1RCvUe6FRE_2-0h4zPLWcUJHDLABFk6hIn6KWHeCxn8F9D__yvxu0Gao8

We talked to LiteSpeed and they said that is was a imagick bug, and reffered us to their solutions:

https://store.litespeedtech.com/store/index.php?rp=%2Fknowledgebase%2F157%2FHTTP-ERROR-when-uploading-images-in-WordPress-with-LiteSpeed.html

However, getting our customers to edit their fuctions.php file is not a great solution. The two other soludtions does not work on their own and also require manual intervention on each client account.

Also imagick is leaving large temp files on our client accounts when i crashes, thereby filling up the space on our client accounts.

So my questions is, if anyone in here have found a host-side fix besides just not using imagick with LiteSpeed webserver?

Thanks!

**Comment by fmw42** [comment]
ImageMagick did not develop nor maintains iMagick.  You should contact the iMagick developer. Crashes can leave ImageMagick temp files in your /tmp directory. But the issue is why is it crashing. It looks like an upload issue. You may want to review and edit your ImageMagick policy.xml files, if you need more memory allocated to ImageMagick. If you fix that, then you will not have those files left around. Nevertheless, it might be a good idea to purge your tmp directory from time to time or even regularly. I believe all ImageMagick temp files will have ImageMagick in the name. So you can purge only those if you desire.

**Reply by MrIverM** [reply]
Hi 

Thank you for your reply, and sorry for the misunderstanding.

As you describe we have a crash issue with upload, but this seems to be related to imagick, since when we uninstall it the problem resolves.

However, we also have the issue with "magick" files in temp folder, which we then though was related to this as well, since customers having theses temp-files only use wordpress.

So is  there a connetion between imagick and ImageMagick? Is ImageMagick utilizing imagick? According to siteground, they seem to be related: https://www.siteground.com/kb/enable-imagick-imagemagick/

Can you point me in the direction of the policy.xml file, then we will give it a try.

**Reply by fmw42** [reply]
iMagick use ImageMagick.  If your temp file do not include the name "ImageMagick", then those are not left from ImageMagick.  The ImageMagick policy information is at https://imagemagick.org/script/resources.php. What I do not understand is that uploading is not part of ImageMagick nor iMagick as I understand it. You have to upload separately from those tools. So that is done in PHP or some other tool. Do I misunderstand your issue?

**Reply by MrIverM** [reply]
Thank you for your reply.

WordPress uses iMagick for processing the images, eg. resizing them - this is what is failing.

I just change memory from 2GiB to 4GiB in policy.xml, but it does not seem to make a differnce.
Do i need to restart someting in orther to have the changes working?

Attached you find the image which is failing. If you have a WP site with a host using imagick and litespeed, you can test yourself.

I just did a test, and when image processing fails, we get a temp file like "magick-dewidj27y7d8h23u".

**Reply by MrIverM** [reply]
This is the image we are testing with: https://imgur.com/a/d5qXCIS

**Reply by MrIverM** [reply]
And error comming from WordPess: Post-processing of the image failed likely because the server is busy or does not have enough resources. Uploading a smaller image may help. Suggested maximum size is 2500 pixels.

**Reply by fmw42** [reply]
Which policy.xml file did you change -- in what directory? Did you change it at `$PREFIX/etc/ImageMagick-7`. If you do not own the server, then the change needs to be done by your hosting organization (or possibly "owner" user of iMagick).  That image is rather large -- 5K x 7K.   If you can increase the memory in your policy.xml file, then it might resolve the issue. Sorry, I do not have Wordpress available and do not use iMagick either. So I do not know if iMagick has any restrictions or control of the policy.xml file. Perhaps one of the ImageMagick or iMagick developers can help you further.

**Reply by MrIverM** [reply]
Hi again,

I edited /etc/ImageMagick-6/policy.xml

We own the server, so we have full access. Even though i make the image smaller eg 200*3000, it still fails - also with 10GiB set in policy.xml

hmmm

**Reply by fmw42** [reply]
What is your ImageMagick version? You tried to change the policy for ImageMagick 6. If you are using ImageMagick 7, then you have to change the policy for that version.  Assuming you have ImageMagick 6, what do you get from 

`convert -version`

and

`convert -list resource`

The latter will tell you what your actual available memory is. Does it match what you changed it to? Does your policy.xml file (resource) allocate enough width and height to permit that 5k x 7k size image? See the resource information you get back.

Do you know what imagick command you were using to process the image at the time it failed?  Do you get any error messages from your PHP code or iMagick?

**Reply by MrIverM** [reply]
From commands:
Version: ImageMagick 6.9.10-68 Q16 x86_64 2020-04-13 https://imagemagick.org
Resources:
  Width: 214.7MP
  Height: 214.7MP
  List length: unlimited
  Area: 269.55GP
  Memory: 125.521GiB
  Map: 251.043GiB
  Disk: unlimited
  File: 3072
  Thread: 64
  Throttle: 0
  Time: unlimited

Changing policy.xml does not seem to change this...so there seems to be someting else dertermining the resources or else my changes are not taking into effect.

**Reply by MrIverM** [reply]
We do not get any php error, other than the one described, and we dont know which imagick command WP use

**Reply by fmw42** [reply]
You have a huge amount of memory -- Memory: 125.521GiB and also width and height of 214.7MP. Those resources should be more than adequate. So I do not know what might be causing the crash.  Do you have more than one version of ImageMagick on the system?  If so, which one is iMagick using?  Might you have shown the resources for a version of ImageMagick that is not being used by iMagick or WordPress?

**Reply by MrIverM** [reply]
Yeah there should be plenty of resources. It only works if we change the dimensions on the image to 1000px in with.

We only have ImageMagick-6 in /etc/, so it seems like this is the only version.

**Reply by fmw42** [reply]
ImageMagick binary is not installed in /etc. You need to find where ImageMagick is installed. Try

`type -a convert`

and

`type -a magick`

See if there is more than one version.

**Reply by MrIverM** [reply]
I get:

convert is /bin/convert
convert is /usr/bin/convert

And 

bash: type: magick: not found

Sorry, I dont know how to interpret the result ;)

**Reply by fmw42** [reply]
You have two versions of ImageMagick 6 installed. That may cause issues.  You can check the versions as follows:

`/bin/convert -version`

and

`/usr/bin/convert -version`

You may have to look at your phpinfo.php file to see which one iMagick is using? Also check to see how WordPress is configured to find imagick or ImageMagick.

**Reply by MrIverM** [reply]
Sorry, i replied below. By the way, this is what LiteSpeed told us:

_It actually is a bug of Imagick only, it tends to leak memory. Most WordPress plugins will use GD in case Imagick is not enabled, the first solution forces WordPress Upload to use GD, even if Imagick is available.

As a host you could disable Imagick at Server-Level, so that all websites can use GD. However, you have to make sure there are no Imagick specific plugins._

**Reply by fmw42** [reply]
So if I understand, this is then not an issue with ImageMagick. It is an issue with iMagick and WordPress.

**Reply by MrIverM** [reply]
I could seem like this.

Thank you for your help, it was worth a try to change policy.xml 👍

**Comment by MrIverM** [comment]
Thanks again - you are being very helpful!

The two versions seems to be the same:

`# /bin/convert -version
Version: ImageMagick 6.9.10-68 Q16 x86_64 2020-04-13 https://imagemagick.org
Copyright: © 1999-2019 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC Modules OpenMP(3.1)
Delegates (built-in): bzlib freetype gslib jng jpeg lcms ltdl lzma openexr png ps tiff wmf x xml zlib
# /usr/bin/convert -version
Version: ImageMagick 6.9.10-68 Q16 x86_64 2020-04-13 https://imagemagick.org
Copyright: © 1999-2019 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC Modules OpenMP(3.1)
Delegates (built-in): bzlib freetype gslib jng jpeg lcms ltdl lzma openexr png ps tiff wmf x xml zlib

I think the first is the main install, and the second is the user-install (we are running a share hosting service)

running phpinfo on our test-site shows:


imagick module version | 3.4.4
imagick classes | Imagick, ImagickDraw, ImagickPixel, ImagickPixelIterator, ImagickKernel
Imagick compiled with ImageMagick version | ImageMagick 6.9.10-68 Q16 x86_64 2020-04-13 https://imagemagick.org
Imagick using ImageMagick library version | ImageMagick 6.9.10-68 Q16 x86_64 2020-04-13 https://imagemagick.org

I dont know if this gives any clues?

**Reply by fmw42** [reply]
You will need to find out which version is used by Imagick and WordPress.  If it your provider shared server, then you cannot change the resources for that one.
